Add basic project.

donghengqaz opened the pull request.

donghengqaz committed:
Add basic project.

donghengqaz commented:
@floitsch PTAL.

floitsch commented on src/esp_box.toit:
```diff
@@ -0,0 +1,109 @@
+
+BSP_CONFIG_DEFAULT ::= bsp_config
+
+class bsp_config:
```
What does "bsp" stand for?
```suggestion
class BspConfig:
```

donghengqaz replied:
The "bsp" is based on C code [here](https://github.com/espressif/esp-box/blob/v0.3.0/components/bsp/src/boards/esp32_s3_box.c#L29).

floitsch replied:
But what does "bsp" mean?
"Board S3 Properties"?

donghengqaz replied:
Yes, because there is another board esp-box-lite, this is a little different from esp-box.

donghengqaz replied:
Actually BSP is "Board Support Package" from some developers in my side.

floitsch replied:
I would have never guessed that name.
Maybe just rename to `BoardConfig` ? It's already in the `esp-box` package, so should be pretty clear.
(And we get into a clash with the "lite" model, we can always rename this one here).

floitsch commented on src/esp_box.toit:
```diff
@@ -0,0 +1,109 @@
+BSP_CONFIG_DEFAULT ::= bsp_config
+
+class bsp_config:
+  I2C_SCL_PIN := 18
```
I think these should be lower case and `::=`.
```suggestion
  i2c_scl_pin ::= 18
```

Same for all other configuration fields.

If you do this, then you should add a constructor that sets the fields. Otherwise it won't be possible to provide your own version.

floitsch commented on src/esp_box.toit:
```diff
@@ -0,0 +1,109 @@
+  TOUCHPAD_I2C_ADDR/int := 0x24
+  TOUCHPAD_I2C_READY_PIN/int := 3
+
+class device:
```
```suggestion
class Device:
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+
+INVALID_VALUE_ ::= -1
+
+bytes2ushort bytes/ByteArray -> int:
```
I would probably use the `LITTLE_ENDIAN` class from `import binary` instead.
See below (the comment on `bytes2ushort data[0..2]`).

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+  button_pressure/int := INVALID_VALUE_
+
+  constructor data/ByteArray:
+    length/int := bytes2ushort data[0..2]
```
```suggestion
    length/int := LITTLE_ENDIAN.uint16 data 0
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+      return "Action is not supported"
+
+class Driver:
+  ADDRESS_/int ::= 0x24
```
I would make this public.
```suggestion
  static I2C_ADDRESS ::= 0x24
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+    else:
+      return "Action is not supported"
+
+class Driver:
```
I would rename to `Tt21100`.
```suggestion
class Tt21100:
```

That said: we haven't applied this consistently. Sometimes we used `Driver` and sometimes we used the name of the sensor. Up to you.

donghengqaz replied:
Yeah, in my opinion I would like also use `Tt21100`, the `Driver` is from `m5stack2` project.

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2023 Toitware ApS.
```
This file should (imho) become its own package. (A `tt21100` package).

I think it's reasonable to commit it here first, and then create a new package and repository with the driver later.

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+  return (bytes[0] + (bytes[1] << 8))
+
+class Action:
+  TOUCH_PAD ::= 0x1
```
```suggestion
  static TOUCH_PAD ::= 0x1
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+
+class Action:
+  TOUCH_PAD ::= 0x1
+  PRESS_BUTTON ::= 0x3
```
```suggestion
  static PRESS_BUTTON ::= 0x3
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+bytes2ushort bytes/ByteArray -> int:
+  return (bytes[0] + (bytes[1] << 8))
+
+class Action:
```
Isn't this more of an `Event` ?

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+TT2100_FLIP_Y  := 0x2
+TT2100_FLIP_XY := 0x4
+
+INVALID_VALUE_ ::= -1
```
The fields in the `Action` class may sometimes end up with `INVALID_VALUE_` in it. For example the `button` might have this set.

As such, we should make this variable public.
```suggestion
INVALID_VALUE ::= -1
```

donghengqaz replied:
The class has `type`, so users can check type and read target fields.

floitsch replied:
Agreed.
I would still make it public, but up to you. It's clearly not a requirement.

floitsch replied:
Never mind. I just saw that you made it public.

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+      button_pressure = bytes2ushort data[6..8]
+
+  /**
+  string format:
```
```suggestion
  Returns a string representation of this instance.
  String format:
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+  ready_pin/gpio.Pin
+  device/i2c.Device
+
+  constructor .width/int .height/int i2c_bus/i2c.Bus .ready_pin/gpio.Pin --frequency/int=400_000 --flags/int=0:
```
I think I would make the width and height named.
Just my preference. It's probably ok to leave them the way they are (since it's extremely common to order them this way).
If other drivers do it without the named argument, leave them unnamed.
```suggestion
  constructor --.width/int --.height/int i2c_bus/i2c.Bus .ready_pin/gpio.Pin --frequency/int=400_000 --flags/int=0:
```

floitsch commented on src/tt21100.toit:
```diff
@@ -0,0 +1,96 @@
+          action.x = action.y
+          action.y = x
+      return action
+    else:
```
I would move this to the top:
```
  if read-pin.get != 0: return null
```
This way we don't need to indent the lines 82-94

floitsch reviewed (approved):
LGTM!
nice.

donghengqaz committed:
Improve code style and writing:

1. make class initial uppercase
2. rename Action to Event
3. make BSP fields lowercase and const
4. use LITTLE_ENDIAN.uint16 to generate short value
5. make some variable public

floitsch commented on src/esp_box.toit:
```diff
@@ -9,101 +9,101 @@ import pixel_display
-  I2C_SDA_PIN := 8
-  I2C_CLOCK := 400_000
+class BspConfig:
+  i2c_scl_pin ::= 18
```
Don't forget to add a constructor so that it's possible to provide a different configuration.

floitsch commented on src/tt21100.toit:
```diff
@@ -64,33 +63,33 @@ class Action:
   device/i2c.Device
 
-  constructor .width/int .height/int i2c_bus/i2c.Bus .ready_pin/gpio.Pin --frequency/int=400_000 --flags/int=0:
+  constructor i2c_bus/i2c.Bus .ready_pin/gpio.Pin --frequency/int=400_000 --flags/int=0 --.width/int --.height/int:
```
Forgot this in my first review.
I think that we usually only take the device and not the bus as argument for drivers.
Not a big change, but would make it more consistent.

floitsch replied:
Also makes it possible to use the same driver for alternative I2C addresses.

floitsch reviewed (approved):
Still LGTM.
(with minor comments).

donghengqaz committed:
Improve code:

1. add constructor to BspConfig
2. tt21100 class uses i2c device instead of i2c bus

donghengqaz committed:
Rename BSPConfig to BoardConfig

floitsch commented:
Thanks!

floitsch commented:
Thanks!

floitsch commented:
Thanks!
